### PR TITLE
Add start, stop and publish a poll for Presenter

### LINKED ISSUE #4888: Start/share a poll

### PR DESCRIPTION
**Pull request by KDSBrowne**
This closes #4888,

remaining items:
~~[ ] Live poll results for Presenter.~~
~~[ ] Update poll response view.~~

### CODE REVIEW

**Comment by Tainan404** [comment]
The poll for presenter isn't updated.

peek 2018-09-25 10-41

**Comment by Tainan404** [comment]
The user interface is different from the one proposed in the @tylercopeland design, in the design a new category is add in the list of users, you add a new option in the action bar.

**Comment by KDSBrowne** [comment]
I thought the proposed design was "a presenter must select the option found under the "+" button.". The image displays a new Polling heading in the userlist. I left this out so I could further discuss it with @tylercopeland

**Comment by tylercopeland** [comment]
@Tainan404 the purposed design does including initiating the poll from the "+" button.  @KDSBrowne I'm happy to chat further about the polling heading in the userlist.

**Comment by oswaldoacauan** [comment]
We will support the new Flash feature where the presenter can see the respondents choice? If yes, we should think about it from start.

**Comment by Tainan404** [comment]
The user can access the polling panel after lose preseter role.
peek 2018-09-25 11-40

File: bigbluebutton-html5/imports/ui/components/actions-bar/container.jsx [hunk 1]
```diff
@@ -5,19 +5,28 @@ import ActionsBar from './component';
...
-  handleUnshareScreen: () => unshareScreen(),
-  isVideoBroadcasting: isVideoBroadcasting(),
-  recordSettingsList: Service.recordSettingsList(),
-  toggleRecording: Service.toggleRecording,
-  screenSharingCheck: getFromUserSettings('enableScreensharing', Meteor.settings.public.kurento.enableScreensharing),
-  enableVideo: getFromUserSettings('enableVideo', Meteor.settings.public.kurento.enableVideo),
-}))(ActionsBarContainer);
+export default withRouter(withTracker(({ location, router }) => {
+  const togglePollMenu = () => {
+    return location.pathname.indexOf('/poll') !== -1
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/actions-bar/container.jsx]: You can use `includes`

File: bigbluebutton-html5/imports/ui/components/poll/component.jsx [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        onChange={() => this.validateInputField(this[`pollInput${i}`])}
+        data-index={i}
+      />);
+    }
+
+    return items;
+  }
+
+  validateInputField(ref) {
+    const option = ref.value.replace(/\s{2,}/g, ' ').trim();
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/poll/component.jsx]: please add one comment explaining what this `regex` do.

File: bigbluebutton-html5/imports/ui/components/button/component.jsx [hunk 3]
```diff
@@ -11,7 +11,7 @@ const SIZES = [
 ];
 
 const COLORS = [
-  'default', 'primary', 'danger', 'success',
+  'default', 'primary', 'danger', 'success', 'primary-outlined',
```
[the remark this answers is not in the document]
Comment (oswaldoacauan) [reply bigbluebutton-html5/imports/ui/components/button/component.jsx]: a ghost button is exactly what you are looking for @KDSBrowne

**Comment by Tainan404** [comment]
When click on `Back to polling options` the state os live polling isn't clean and the polling response container isn't removed when back to polling option.

peek 2018-10-16 13-32

**Comment by Tainan404** [comment]
The polling container overrides the dropdowns, is this expected?

screenshot from 2018-10-16 13-35-40

**Comment by Tainan404** [comment]
it would not be better to change the circumflex accent to something more intuitive like an "unanswered" string?
screenshot from 2018-10-16 13-43-03

**Comment by Tainan404** [comment]
The live poll shows the offline users:

peek 2018-10-25 14-05

File: bigbluebutton-html5/imports/ui/components/chat/message-list/message-list-item/styles.scss [hunk 4]
```diff
@@ -17,6 +17,7 @@
 
 .systemMessage {
   padding: var(--line-height-computed) 0;
+  padding-top: 0px;
```
Comment (antobinary) [inline bigbluebutton-html5/imports/ui/components/chat/message-list/message-list-item/styles.scss]: I think the `0` is included in the above line. No need of padding-top here

**Comment by Tainan404** [comment]
The polling title is misaligned and the polling button doesn't toggling.

peek 2018-10-26 09-09

**Comment by KDSBrowne** [comment]
Is it suppose to toggle?  Or should it behave like the Chat buttons?

**Comment by Tainan404** [comment]
Yes

**Comment by Tainan404** [comment]
When hide the polling panel the polling is closed. The polling Container would shouldn't be positioned on the left-bottom, overlapping the video dock? If it changed to center make that with one modal

peek 2018-10-29 09-37

File: bigbluebutton-html5/imports/ui/components/poll/live-result/container.jsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import Users from '/imports/api/users';
+import Polls from '/imports/api/polls';
+import Auth from '/imports/ui/services/auth';
+import { withTracker } from 'meteor/react-meteor-data';
+import LiveResult from './component';
+
+const LiveResultContainer = ({ ...props }) => <LiveResult {...props} />;
+
+export default withTracker(({ }) => {
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/container.jsx]: You don't need a object destruction here

File: bigbluebutton-html5/imports/ui/components/actions-bar/container.jsx [hunk 6]
```diff
@@ -8,16 +9,35 @@ import { shareScreen, unshareScreen, isVideoBroadcasting } from '../screenshare/
...
+
+    const show = () => {
+      Session.set('isUserListOpen', true);
+      Session.set('isPollOpen', true);
+      Session.set('forcePollOpen', true);
+    };
+
+    const hide = () => Session.set('isPollOpen', false);
+
+    if (Session.equals('isChatOpen', true)) Session.set('isChatOpen', false);
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/actions-bar/container.jsx]: Are you sure you want this sideeffect?

We probably want to have something more robust and flexible regarding on what is "open" in the sidebar. When we merge #6105 and #6178 we will have 3 possible states: Chat, Breakouts and Polls, otherwise we will endup with one session variable for each state and start to create eadge caes like: what happens if all of them are "open"? Is that a possible state?

What do you think @antobinary?

File: bigbluebutton-html5/imports/ui/components/poll/component.jsx [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+          label={intl.formatMessage(intlMessages.startCustomLabel)}
+          color="primary"
+          aria-disabled={isDisabled}
+          className={styles.btn}
+        />
+      </div>
+    );
+  }
+
+  back() {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/component.jsx]: `handleBackClick`

File: bigbluebutton-html5/imports/ui/components/poll/component.jsx [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+      return Session.set('isPollOpen', false);
+    }
+  }
+
+  toggleCustomFields() {
+    const { customPollReq } = this.state;
+
+    this.inputEditor = [];
+
+    return customPollReq
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/component.jsx]: `this.setState({ customPollReq: !this.state.customPollReq })`

File: bigbluebutton-html5/imports/ui/components/poll/component.jsx [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+
+  handleInputChange(index, event) {
+    // This regex will replace any instance of 2 or more consecutive white spaces
+    // with a single white space character.
+    const option = event.target.value.replace(/\s{2,}/g, ' ').trim();
+    this.inputEditor[index] = option === '' ? '' : option;
+    this.setState({ customPollValues: this.inputEditor });
+  }
+
+  getInputFields() {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/component.jsx]: `renderInputFields`

File: bigbluebutton-html5/imports/ui/components/poll/component.jsx [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+    this.handleInputChange = this.handleInputChange.bind(this);
+    this.back = this.back.bind(this);
+  }
+
+  componentWillMount() {
+    this.nonPresenterRedirect();
+  }
+
+  componentDidUpdate() {
+    this.nonPresenterRedirect();
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/component.jsx]: This probably should be a `HoC` since there is more components that would have this same behavior. cc @Tainan404 

Both PR`s handle the same use-case differently.

File: bigbluebutton-html5/imports/ui/components/poll/container.jsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+export default withTracker(({ }) => {
+  Meteor.subscribe('current-poll', Auth.meetingID);
+
+  const currentPresentation = Presentations.findOne({
+    current: true,
+  });
+
+  const currentSlide = PresentationAreaService.getCurrentSlide(currentPresentation.podId);
+
+  const currentUser = Users.findOne({ userId: Auth.userID });
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/container.jsx]: You could probably move most of those to a `service.js` like most of the big containers do.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+class LiveResult extends Component {
+  constructor(props) {
+    super(props);
+
+    this.getUnresponsive = this.getUnresponsive.bind(this);
+    this.getRespondents = this.getRespondents.bind(this);
+    this.getPollStats = this.getPollStats.bind(this);
+  }
+
+  getPollStats() {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: `renderPollStats`

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+            </div>
+          </div>);
+        });
+      }
+    }
+
+    return pollStats;
+  }
+
+  getRespondents() {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: `renderRespondents`

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    const { currentPoll, getUser } = this.props;
+
+    if (!currentPoll) return null;
+
+    const respondedUsers = [];
+
+    if (currentPoll) {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: You already return null in L73 if there is not poll.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+    const pollStats = [];
+
+    if (currentPoll) {
+      const {
+        answers,
+        numRespondents,
+      } = currentPoll;
+
+      if (answers) {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: You could return early.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+    this.getRespondents = this.getRespondents.bind(this);
+    this.getPollStats = this.getPollStats.bind(this);
+  }
+
+  getPollStats() {
+    const { currentPoll } = this.props;
+
+    const pollStats = [];
+
+    if (currentPoll) {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: You could return early.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+
+    const {
+      users,
+    } = currentPoll;
+
+    const usersToRespond = [];
+
+    const usersList = _.compact(users);
+
+    if (usersList) {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: You could return early.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+      usersList.map((userId) => {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: You are using `map` as a `forEach`.

```js
usersList
  .map((userId) => getUser(userId))
  .filter(user => user && user.connectionStatus == 'online')
  .reduce((acc, user) => [
    ...acc,
    <div className={styles.item} key={_.uniqueId('stats-')}>{user.name}</div>,
    <div className={styles.itemR} key={_.uniqueId('stats-')}>-</div>,
  ], [])
```

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+        const user = getUser(userId);
+
+        if (user && user.connectionStatus == 'online') {
+          usersToRespond.push(<div className={styles.item} key={_.uniqueId('stats-')}>{user.name}</div>);
+          usersToRespond.push(<div className={styles.itemR} key={_.uniqueId('stats-')}>-</div>);
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: You could do the `-` using css `:after` and just hiding it on the `:last-child`.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+      const {
+        answers,
+        responses,
+      } = currentPoll;
+
+      if (responses && answers) {
+        responses.map((ur) => {
+          const user = getUser(ur.userId);
+          if (user) {
+            answers.map((obj) => {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: Same as the review bellow, you are using map as forEach.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+            });
+          }
+        });
+      }
+    }
+
+    return respondedUsers;
+  }
+
+  getUnresponsive() {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/component.jsx]: Both `getRespondents` and `getUnresponsive` you probably could reuse some code between the two or create a render function/component that receives an array of userIds and render them as you want.

File: bigbluebutton-html5/imports/ui/components/poll/live-result/styles.scss [hunk 22]
```diff
@@ -0,0 +1,110 @@
+@import "/imports/ui/stylesheets/variables/_all";
+
+:root {
+  --stats-border-color: #d4d9df;
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/live-result/styles.scss]: I would suggest you to prefix those variables with `poll-` since they are global.

File: bigbluebutton-html5/imports/ui/components/poll/styles.scss [hunk 23]
```diff
@@ -0,0 +1,166 @@
+@import "/imports/ui/stylesheets/variables/_all";
+@import "/imports/ui/stylesheets/mixins/focus";
+
+:root {
+    --column-amount: 2;
+    --poll-blue: #1A73D4;
+    --headerOffset: -0.875rem;
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/poll/styles.scss]: kebab-case

Comment (antobinary) [reply bigbluebutton-html5/imports/ui/components/poll/styles.scss]: souvlaki-case :flag-gr:

Comment (oswaldoacauan) [reply bigbluebutton-html5/imports/ui/components/poll/styles.scss]: pacoca-case :brazil:

**Comment by Tainan404** [comment]
peek 2018-10-31 14-10

**Comment by Tainan404** [comment]
screenshot from 2018-11-05 15-35-42

File: bigbluebutton-html5/imports/ui/components/whiteboard/service.js [hunk 24]
```diff
@@ -32,19 +32,23 @@ function handleAddedAnnotation({
...
-        position: annotation.position,
-        'annotationInfo.color': isEqual(fakePoints, lastPoints) || annotation.status === DRAW_END ?
-          annotation.annotationInfo.color : fakeAnnotation.annotationInfo.color,
-      },
-      $inc: { version: 1 }, // TODO: Remove all this version stuff
-    });
-    return;
+  let fakePoints;
+
+  if (fakeAnnotation) {
```
Comment (oswaldoacauan) [inline bigbluebutton-html5/imports/ui/components/whiteboard/service.js]: Is this related to this PR?

Comment (KDSBrowne) [reply bigbluebutton-html5/imports/ui/components/whiteboard/service.js]: Yes, If I remember correctly is preventing an error from 'annotationInfo' of undefined when the poll results are published.
